Add shell script for downloading JAR dependencies

nre-ableton opened the pull request:
ping @AbletonDevTools/gotham-city

mst-ableton commented:
I hate to be the person to ask for a rewrite but, why is this a shell script? It seems like the script is doing complicated enough parsing that we should make it a small Python script that calls `os.system`. It might be possible to do something with `subprocess` that would be 'cleaner,' but I believe you would have to work harder to get nice TTY output.

ala-ableton commented:
Please don't use `os.system`. Use `requests` for downloading, and `tarfile` and `gzip` for extracting.

nre-ableton commented:
@AbletonDevTools/gotham-city I've rewritten this change and now the `fetch_jars` script is in Python (which is admittedly kinda nice, since the output from `requests` is better than `wget`). Mind taking another look?

nre-ableton commented:
@ala-ableton Whoops, I just noticed that I didn't use the `tarfile` module as you'd recommended. Fixup coming.

ala-ableton commented:
@nre-ableton no problem. Please also reword the commit message :wink:

nre-ableton commented:
@ala-ableton Done! Also rebased/squashed.

mst-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,121 @@
+
+    if os.path.exists(output_file_path):
+        if force:
+            os.remove(output_file_path)
```
_technically_ we should trap the file does not exist error here in case there was a race condition between `os.path.exists` but I don't mind skipping it here

ala-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,121 @@
+    # Uncompress SLF4J tarball
+    slf4j_path = os.path.join(args.output_dir, slf4j_file_name)
+    logging.debug('Uncompressing %s', slf4j_path)
+    tar = tarfile.open(slf4j_path)
```
Please use a context manager here:
```python
with tarfile.open(slf4j_path) as tar:
    ...
```

ala-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,121 @@
+    arg_parser = argparse.ArgumentParser()
+
+    arg_parser.add_argument(
+        '--codenarc',
```
`--codenarc-version` would be nicer. You can still type `--codenarc` (or even `--co`) when calling this script, because argparse will use the argument that matches.

ala-ableton replied:
`--codenarc-version` would be nicer. You can still type `--codenarc` (or even `--co`) when calling this script, because argparse will use the argument that matches.

ala-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,121 @@
+    )
+
+    arg_parser.add_argument(
+        '--gmetrics',
```
`--gmetrics-version` would be nicer. You can still type `--gmetrics` (or even `--gm`) when calling this script, because argparse will use the argument that matches.

ala-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,121 @@
+    )
+
+    arg_parser.add_argument(
+        '--slf4j',
```
`--slf4j-version` would be nicer. You can still type `--slf4j` (or even `--sl`) when calling this script, because argparse will use the argument that matches.

nre-ableton commented:
@mst-ableton ah, sorry, I misread the diff. I was on my phone at the time. 😅

nre-ableton commented:
@ala-ableton thanks for the review, I think I’ve addressed your comments. I’ve also rebased/squashed @mst-ableton‘s changes and force pushed. Mind taking another look?

ala-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,121 @@
+
+    jar_urls = [
+        'https://netcologne.dl.sourceforge.net/project/codenarc/codenarc'
+        f'/CodeNarc%20{args.codenarc_version}/CodeNarc-{args.codenarc_version}.jar',
```
Please put parens around these two lines to highlight that it is a single string.

nre-ableton replied:
Do you mean like:

```python
(
    ‘blah’
    ‘blah’
),
```

Sorry for my python naivety.

rco-ableton replied:
Answering for @ala-ableton: yes 🙂

ala-ableton replied:
I meant something like this
```python
    jar_urls = [
        ('https://netcologne.dl.sourceforge.net/project/codenarc/codenarc'
         f'/CodeNarc%20{args.codenarc_version}/CodeNarc-{args.codenarc_version}.jar'),
        ('https://github.com/dx42/gmetrics/releases/download'
         f'/v{args.gmetrics_version}/GMetrics-{args.gmetrics_version}.jar'),
        f'https://www.slf4j.org/dist/{slf4j_file_name}',
    ]
```
I wouldn't add another level of indentation inside the parens, because it could make the lines too long, thus defeating the point of adding the parens :smile:

nre-ableton replied:
@ala-ableton Thanks for the clarification!

ala-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,121 @@
+        'https://netcologne.dl.sourceforge.net/project/codenarc/codenarc'
+        f'/CodeNarc%20{args.codenarc_version}/CodeNarc-{args.codenarc_version}.jar',
+        'https://github.com/dx42/gmetrics/releases/download'
+        f'/v{args.gmetrics_version}/GMetrics-{args.gmetrics_version}.jar',
```
Please put parens around these two lines to highlight that it is a single string.

ala-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,121 @@
+    logging.debug('Uncompressing %s', slf4j_path)
+    with tarfile.open(slf4j_path) as tar:
+        tar.extractall(path=args.output_dir)
+        tar.close()
```
The context manager will take care of this. Don't close twice!

ala-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,121 @@
+    with tarfile.open(slf4j_path) as tar:
+        tar.extractall(path=args.output_dir)
+        tar.close()
+        logging.info('Uncompressed SLF4J')
```
Please unindent this statement out of the context manager. It will match line 37.

ala-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,121 @@
+        help='Show verbose output.',
+    )
+
+    args = arg_parser.parse_args(sys.argv[1:])
```
You don't need to pass `sys.argv[1:]`, that's the default.

rco-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,120 @@
+
+
+def download_file(url, output_dir, force=False):
+    """Download a file from the internet to the download directory."""
```
`from a URL` might be more technically correct?

rco-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,120 @@
+    output_file_name = url.split('/')[-1]
+    output_file_path = os.path.join(output_dir, output_file_name)
+
+    if os.path.exists(output_file_path):
```
It would be more Pythonic to try the removal and trap the `IOError` than to check first (especially for the reason @mst-ableton already mentioned).

ala-ableton replied:
@rco-ableton I think you mean the `FileNotFoundError` exception :wink:

rco-ableton replied:
@ala-ableton I actually meant `OSError`, since there might be other exceptions raised 😉

ala-ableton replied:
Then I disagree with you. If there is another kind of error, which is not due to the file not existing, we shouldn't trap it.

rco-ableton replied:
That's a good point. So yes, `FileNotFoundError` should be caught and other exceptions can get thrown up the stack.

rco-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,120 @@
+            return
+
+    logging.debug('Downloading %s to %s', url, output_file_path)
+    response = requests.get(url)
```
What happens if the response is an error?

nre-ableton replied:
Annoyingly enough, nothing. So in the case of sourceforge, when I give it an intentionally bad URL, I get redirected to a valid page (so in other words, a `200` status code).

Maybe using `requests` is a bad idea and we should just exec out to `wget` as we were doing before. 🤔

nre-ableton replied:
Hmm nevermind, I think I have an idea as to how we can get around this...

rco-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,120 @@
+
+    slf4j_file_name = f'slf4j-{args.slf4j_version}.tar.gz'
+
+    jar_urls = [
```
This block of code is a bit messy and hard to read (not your fault—just because of the horrible URLs and concatenation). I wonder if it might make sense to read the URLs from an external file, where we can safely put each URL on its own line and not have to worry about string interpolation etc?

rco-ableton replied:
Oh, on second thoughts I see you'd still have to do some templating. Still, maybe it would still work to extract the templates to a file? Or to a set of globals? I'm not sure which would look better.

nre-ableton replied:
I agree that the present solution is a bit ugly, I don't want to run the risk of over-engineering what should be a very simple downloading script (at least not in the initial rewrite). I would be happy to make an issue for how we can improve this in the future, though. Sound ok?

rco-ableton replied:
Sure.

ala-ableton commented on fetch_jars.py:
```diff
@@ -17,20 +17,24 @@
     logging.debug('Downloading %s to %s', url, output_file_path)
     response = requests.get(url)
 
+    if not (200 <= response.status_code < 300):
```
I think we should prefer `response.ok`.

nre-ableton replied:
Thanks, I didn’t know about this.

ala-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,124 @@
+    output_file_name = url.split('/')[-1]
+    output_file_path = os.path.join(output_dir, output_file_name)
+
+    if force:
```
So if `force=False`, we don't delete the maybe existing file, but we still download? I don't understand why it's called `force` then...

nre-ableton replied:
Hmm no, if force is false, then we should return if the file exists. IMHO the previous version of this code was nicer...

rco-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,128 @@
+    slf4j_file_name = f'slf4j-{args.slf4j_version}.tar.gz'
+
+    jar_urls = [
+        ('https://netcologne.dl.sourceforge.net/project/codenarc/codenarc'
```
Style nitpick: the parentheses go on their own lines (this is technically "list formatting" as per our draft style guide, but with the exception of no trailing commas):

```python
(
    'https://netcologne.dl.sourceforge.net/project/codenarc/codenarc'
    f'/CodeNarc%20{args.codenarc_version}/CodeNarc-{args.codenarc_version}.jar'
),
```

nre-ableton replied:
That’s what I thought. 😿 ping @ala-ableton

ala-ableton replied:
I don't consider this list formatting, since these are not items and there are no commas. However, I don't mind using that style, since it seems to fit in the line width.

nre-ableton replied:
Yeah, I guess also we will (hopefully) move these URLs to a separate templated file or something. So I'm also fine with this style for now.

ala-ableton commented on fetch_jars.py:
```diff
@@ -26,6 +26,10 @@ def download_file(url, output_dir, force=False):
         except FileNotFoundError:
             pass
+    else:
+        if os.pah.exists(output_file_path):
```
Could be `elif`.

nre-ableton replied:
True, that’s a bit nicer.

rco-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,128 @@
+        raise ValueError(f'Expected raw data for {url}, but got {response.encoding} data')
+
+    with open(output_file_path, mode='wb') as output_file:
+        output_file.write(response.content)
```
You should prefer using `response.text` over `response.content` as the former will decode the content according to the encoding specified by the server.

nre-ableton replied:
But we check above that the encoding is `None`. We definitely don’t want to convert the content type, since we are expecting to receive binary data.

rco-ableton replied:
Ah, in that case you might want to look at [using `requests` to fetch binary data](https://requests.readthedocs.io/en/latest/user/quickstart/#raw-response-content).

rco-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,128 @@
+
+    if not response.ok:
+        raise ValueError(f'Failed to fetch {url}')
+    if response.encoding is not None:
```
What's the purpose of this check?

nre-ableton replied:
Is the error message not clear enough here? I’m not being sarcastic, just asking honestly.

rco-ableton replied:
Sorry, I phrased my question badly. Is this essentially nothing more than an error handling check?

nre-ableton replied:
More or less. It's dealing with the stupidity of SourceForge, but actually, I guess we could remove it. So for a URL like `https://netcologne.dl.sourceforge.net/project/codenarcxxxx/etc`, SourceForge redirects us to `/create` and we get an HTML page (and status code 200) with a bunch of crap on it. However, for `https://netcologne.dl.sourceforge.net/project/codenarc/codenarcxxx/etc`, we get a sensible 404.

So the check is basically here to make sure we are being served a binary file and not HTML text content. We could remove this, but then `run_codenarc.py` will probably fail with a very odd error like `failed to load org.codenarc.CodeNarc from classpath`.

rco-ableton replied:
How does this work if the server returns the `Content-Type` as `application/octet-stream` or some other binary type?

nre-ableton replied:
@rco-ableton I'm not sure how `requests` will deal with that. For the 3 URLs in our script, the `encoding` is `None`. I'm pretty sure that this is not the same as the content type, and the documentation for this field states:

> Encoding to decode with when accessing r.text.

Anyways, I'm fine removing this check for now, since it's starting to feel that this check is depending on the internal behavior of the framework, which is not a great sign. Anyways, I think I have another solution here, fixup is coming....

rco-ableton replied:
I had a brief look through the source code and I _think_ that `encoding` will remain `None` unless the `Content-Type` is `text/something`. I wouldn't want to guarantee that though.

I'm not sure about https://github.com/AbletonDevTools/groovylint/commit/32fcf6077af00e4d9531d01f49285d2845fecad9, only because having a list of whitelisted MIME types still seems liable to break in unexpected ways if/when we add more URLs to the list.

I wonder if it makes sense to download the binary data as-is and then actually error-check the downloaded file (e.g. by trying to extract it in some way and catching any errors there)?

nre-ableton replied:
Maybe. A JAR file is just a zipfile after all, we could try to open it somehow with the `tarfile` module, which would fail if we get text. I'll give that a shot...

nre-ableton replied:
@rco-ableton Have a look at 8805053. This is starting to feel much better. `zipfile` will raise `raise BadZipFile("File is not a zip file")` if it can't open the file, and if we get a real zipfile which isn't an actual JAR, we can raise our own exception (`MANIFEST.MF` is the only required file in a JAR, IIRC).

Note that for the tarballs, we'll get a separate exception if that content is not what we are expecting. What do you think of this approach?

rco-ableton replied:
This looks reasonable to me. I'd be tempted to extract checking downloaded binaries to a separate function that determines the right procedure to verify that they are what they claim to be, and then maybe refactoring the downloading into a function too:

```python
def _verify(file):
    if file.endswith('.jar'):
        _verify_is_jar(file)
    elif file.endswith('.gz'):
        _verify_is_tarfile(file)

def _download_and_verify(file):
    r = requests.get(file)
   
    if r.ok:
        _verify(file)

...

for file in jar_files:
   _download_and_verify(file)
```

etc.

nre-ableton replied:
@rco-ableton I agree that verification should be split off into a separate function. However, I don't like the idea of verifying the tarballs. That means we'll basically have 2 duplicate lines with `with tarfile.open(...)`. How about this:

```python
    for url in jar_urls:
        file_type = download_file(url, args.output_dir, args.force)
        if file_type == 'jar':
            verify_jar_file(output_path)
        elif file_type == 'tgz':
            extract_tarball(output_path)
```

rco-ableton replied:
I'm not sure about returning the file type from `download_file()`—that seems surprising to me, just from an API-design point of view. Are we expecting the tarballs to contain jarfiles themselves?

nre-ableton replied:
@rco-ableton The SLF4J tarball contains a bunch of stuff, mostly JAR files, but also some text files, a `pom.xml`, etc.

rco-ableton replied:
Right, shame. I was wondering if we could re-use some logic once we'd extracted the tarball. I guess we could go with what you suggest, given that I think we're mostly bike-shedding now. Input from @ala-ableton and @mst-ableton would be good here, too.

ala-ableton replied:
My take on this:
```python
def download_file(url, output_dir, force=False):
    """Download a file from a URL to the download directory."""
    output_file_name = url.split('/')[-1]
    output_file_path = os.path.join(output_dir, output_file_name)

    if force:
        try:
            os.remove(output_file_path)
        except FileNotFoundError:
            pass
    elif os.path.exists(output_file_path):
        logging.debug('%s already exists, skipping download', output_file_path)
        return output_file_path

    logging.debug('Downloading %s to %s', url, output_file_path)
    response = requests.get(url, stream=True)

    if not response.ok:
        raise ValueError(f'Failed to fetch {url}')

    with open(output_file_path, mode='wb') as output_file:
        for chunk in response.iter_content(chunk_size=256):
            output_file.write(chunk)

    logging.info('Downloaded %s', output_file_name)
    return output_file_path


def verify_jar(file_path):
    """Verify that ``file_path`` is a valid JAR file."""
    logging.debug('Verifying %s', file_path)
    with zipfile.ZipFile(file_path, 'r') as jar_file:
        if 'META-INF/MANIFEST.MF' not in jar_file.namelist():
            raise ValueError(f'{file_path} does not appear to be a valid JAR')


def fetch_jars(args):
    """Fetch JAR file dependencies."""
    if not os.path.exists(args.output_dir):
        os.mkdir(args.output_dir)

    codenarc_jar_path = download_file(
        ('https://netcologne.dl.sourceforge.net/project/codenarc/codenarc'
         f'/CodeNarc%20{args.codenarc_version}/CodeNarc-{args.codenarc_version}.jar'),
        args.output_dir,
        args.force,
    )
    verify_jar(codenarc_jar_path)

    gmetrics_jar_path = download_file(
        ('https://github.com/dx42/gmetrics/releases/download'
            f'/v{args.gmetrics_version}/GMetrics-{args.gmetrics_version}.jar'),
        args.output_dir,
        args.force,
    )
    verify_jar(gmetrics_jar_path)

    slf4j_tar_path = download_file(
        f'https://www.slf4j.org/dist/slf4j-{args.slf4j_version}.tar.gz',
        args.output_dir,
        args.force,
    )
    # Uncompress SLF4J tarball
    logging.debug('Uncompressing %s', slf4j_tar_path)
    with tarfile.open(slf4j_tar_path) as tar:
        tar.extractall(path=args.output_dir)
    logging.info('Uncompressed SLF4J')

    return 0
```

nre-ableton replied:
@ala-ableton thanks for the suggestion. I decided to go with a similar but slightly different solution. Mind taking another look? Also ping @rco-ableton

ala-ableton replied:
@nre-ableton very nice :clap:

ala-ableton commented on fetch_jars.py:
```diff
@@ -60,11 +60,13 @@ def fetch_jars(args):
-        f'https://www.slf4j.org/dist/{slf4j_file_name}',
+        (
+            f'https://www.slf4j.org/dist/{slf4j_file_name}',
+        ),
```
Why do we end-up with tuples?

nre-ableton replied:
Because surrounding the multi-line URLs with parenthesis makes them tuples, and passing `url[0]` below fixes that but then results in `h` for the SLF4J one.

ala-ableton replied:
> Because surrounding the multi-line URLs with parenthesis makes them tuples

No. What makes them tuples is putting commas within the parens. As I said earlier, there should not be commas within the parens.

nre-ableton replied:
Ah sorry, I misunderstood your above comment. My bad!

ala-ableton commented:
@nre-ableton once rebased, LGTM :+1:

nre-ableton commented:
@rco-ableton Any last opinions on this PR before I merge?

rco-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,145 @@
+    for url in tar_urls:
+        uncompress_tar(download_file(url, args.output_dir, args.force), args.output_dir)
+
+    return 0
```
This is super-nitpicky but whenever I see integer return codes in Python I shudder slightly. The Pythonic way, in my opinion, is to raise exceptions if something goes wrong. If not, then we don't need to return anything. Any exceptions raised here will cause a non-zero exit code from the entry point anyway.

If you're worried about the exception being dumped to the console, I think the more correct thing to do would be to catch it within `if __name__ == '__main__'` and `sys.exit()` appropriately.

nre-ableton replied:
I'm fine with:

```python
if __name__ == '__main__':
    fetch_jars(parse_args())
    sys.exit(0)
```

If `fetch_jars` raises, then having an exception barf all over the console is fine. I've force-pushed my fix.

rco-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,144 @@
+
+if __name__ == '__main__':
+    fetch_jars(parse_args())
+    sys.exit(0)
```
This line doesn't do anything—the default return code from the interpreter is `0`.

nre-ableton replied:
Good point, thanks!

rco-ableton commented on fetch_jars.py:
```diff
@@ -0,0 +1,144 @@
+    """Uncompress a tarball to the given output directory."""
+    logging.debug('Uncompressing %s', file_path)
+    with tarfile.open(file_path) as tar:
+        tar.extractall(path=output_dir)
```
Whilst reading the docs for `tarfile` I came across this:

> **Warning**: Never extract archives from untrusted sources without prior inspection. It is possible that files are created outside of path, e.g. members that have absolute filenames starting with "/" or filenames with two dots "..".

What's our line of thinking on this? I guess we will have to extract them at some point, but maybe we should be building some logic to make sure they're not malicious tarballs (i.e. don't contain paths above the extraction target)?

nre-ableton replied:
That's a fair point. But also, we've gone from 4 lines in a `Dockerfile` to a ~150 line Python script, and IMHO we are risking over-engineering this task. 😅

I'm happy to make an issue for this so it's not forgotten, but given that we're not downloading arbitrary tarballs I think the risk is pretty minimal.

rco-ableton replied:
Indeed! I don't think we necessarily need to make an issue to do this in the future (I'm not stopping you, however). What about a comment above `tar_urls` explaining that you should be careful when adding to this list, referencing the above warning?

mst-ableton replied:
> building some logic to make sure they're not malicious tarballs

I recommend a different solution. Let's mirror both JARs here, on GitHub.

rco-ableton replied:
> I recommend a different solution. Let's mirror both JARs here, on GitHub.

Binary files, in Git?! 😅

(It's also a reasonable solution, but doesn't help if someone else who uses this script wants to add more tarballs without understanding the risks.)

mst-ableton replied:
@nre-ableton points out that this adds complexity to making releases.
